maskrcnn_benchmark/structures/segmentation_mask.py

```diff
@@ -1,51 +1,92 @@
+                    h, w = segm['size']
+                    rle = mask_utils.frPyObjects(segm, h, w)
+                    mask = mask_utils.decode(rle)
+                    mask = torch.from_numpy(mask).to(dtype=torch.float32)
+                else:
+                    # compressed RLE
+                    mask = mask_utils.decode(segm)
+                    mask = torch.from_numpy(mask).to(dtype=torch.float32)
+            else:
+                # binary mask
+                if type(segm) == np.ndarray:
+                    mask = torch.from_numpy(segm).to(dtype=torch.float32)
+                else: # torch.Tensor
+                    mask = segm.to(dtype=torch.float32)
+        self.mask = mask
         self.size = size
         self.mode = mode
 
     def transpose(self, method):
         if method not in (FLIP_LEFT_RIGHT, FLIP_TOP_BOTTOM):
-            raise NotImplementedError(
-                "Only FLIP_LEFT_RIGHT and FLIP_TOP_BOTTOM implemented"
-            )
+            raise NotImplementedError("Only FLIP_LEFT_RIGHT and FLIP_TOP_BOTTOM implemented")
 
         width, height = self.size
         if method == FLIP_LEFT_RIGHT:
-            dim = width
-            idx = 2
+            max_idx = width
+            dim = 1
         elif method == FLIP_TOP_BOTTOM:
-            dim = height
-            idx = 1
+            max_idx = height
+            dim = 0
 
-        flip_idx = list(range(dim)[::-1])
-        flipped_masks = self.masks.index_select(dim, flip_idx)
-        return Mask(flipped_masks, self.size, self.mode)
+        flip_idx = torch.tensor(list(range(max_idx)[::-1]))
+        flipped_mask = self.mask.index_select(dim, flip_idx)
+        return Mask(flipped_mask, self.size, self.mode)
 
     def crop(self, box):
-        w, h = box[2] - box[0], box[3] - box[1]
-
-        cropped_masks = self.masks[:, box[1] : box[3], box[0] : box[2]]
-        return Mask(cropped_masks, size=(w, h), mode=self.mode)
+        box = [round(b) for b in box]
+        w, h = box[2] - box[0] + 1, box[3] - box[1] + 1
+        w = max(w, 1)
+        h = max(h, 1)
+        cropped_mask = self.mask[box[1]: box[3], box[0]: box[2]]
+        return Mask(cropped_mask, size=(w, h), mode=self.mode)
 
     def resize(self, size, *args, **kwargs):
-        pass
+        width, height = size
+        scaled_mask = interpolate(self.mask[None, None, :, :], (height, width), mode='bilinear')[0, 0]
+        return Mask(scaled_mask, size=size, mode=self.mode)
+
+    def convert(self, mode):
+        mask = self.mask.to(dtype=torch.uint8)
+        return mask
+
+    def __iter__(self):
+        return iter(self.mask)
+
+    def __repr__(self):
+        s = self.__class__.__name__ + "("
+        # s += "num_mask={}, ".format(len(self.mask))
+        s += "image_width={}, ".format(self.size[0])
+        s += "image_height={}, ".format(self.size[1])
+        s += "mode={})".format(self.mode)
+        return s
+
 
 
 class Polygons(object):
```

Comment:
I think the `mode` field for the Polygon is completely irrelevant.
It is never used, but causes:
-  additional argument passing when constructing
- a headache when try to find out what Polygon really is

**Question:**
Wouldn't it be more consistent if the `convert` method of a Polygon would be renamed to `convert_to_mask` and would return a Mask instance?

Reply:
Hi,

I agree with your points.

The reason why this is currently the case was that I wanted to keep the same interface between `Polygons` and `Box` (which is not implemented, but is the single-box equivalent of `BoxList`).
And my original idea was that we would be able to specify what was the underlying type of the data via the mode: is it a polygon, or a mask?

I'm not sure about changing the `convert` name of the method though.

In general, I think both `box_list` and `segmentation_mask` could benefit from some better design / cleanup, but I'm not sure what that would be

Reply:
> And my original idea was that we would be able to specify what was the underlying type of the data via the mode: is it a polygon, or a mask?

I think I cannot follow this part:
> To specify what was the underlying type of the data via the mode

As in the current implementation, a `Polygon` instance: 
1.  can be initialized either with a `list` of polygons
1.  can be initialized either with a `Polygon` instance (which is referenced now, but **should** be hard-copied IMO)
1.  **cannot** be initialized with a `Mask`, which feature could be added if necessary (I am doing this to convert GTA binary masks to COCO Polygon format, but only because the binary masks are not supported).

So the underlying data would be specified: Polygon.

On the other hand, about the `convert` function:
>I'm not sure about changing the `convert` name of the method though.
1. The `convert` function takes an argument for the target mode, but actually it accepts just a single answer, which is odd.
2. If I assume that a `Polygon` can be only `convert`-ed to a `Mask` than `convert` name is OK, but relies on the assumption that the data can be represented either in `Polygon`s or `Mask`s and nothing else, which is not necessary a trivial assumption, so changing the name to `convert_to_mask` would be clear from the very first encounter.

> keep the same interface
1. We should in this case add the `convert` or `convert_to_polygon` method to the `Mask` class as well

Reply:
Those are all reasonable points, and I'm willing to accept PRs that improve the overall consistency and software design of the codebase

Reply:
@fmassa Thanks, these points were considered for the refactored version, PR #473